_posts/2018-05-21-shipped-and-hosted-software.md

```diff
@@ -0,0 +1,268 @@
+[Kubernetes](http://kubernetes.io/) resources for deployment with `kubectl`.
+Kubernetes and Terraform both have limited support for migration from one
+release to another: Terraform will only create or modify changed resources, and
+Kubernetes will perform a phased roll-out of any modified resources.
+
+By the way, all of this build-and-release functionality is implemented in the
+new [taskcluster-installer](https://github.com/taskcluster/taskcluster-installer).
+
+### Service Discovery
+
+The string `taskcluster.net` appears quite frequently in the Taskcluster source
+code. For any other deployment, that hostname is not valid -- but how will the
+service find the correct hostname? The question extends to determining pulse
+exchange names, task artifact hostnames, and so on. There are also security
+issues to consider: misconfiguration of URLs might enable XSS and CSRF attacks
+from untrusted content such as task artifacts.
+
+The approach we are taking is to define a `rootUrl` from which all other URLs
+and service identities can be determined. Some are determined by simple
+transformations encapsulated in a new
+[taskcluster-lib-urls](https://github.com/taskcluster/taskcluster-lib-urls)
+library. Others are fetched at runtime from other services: pulse exchanges
+from the taskcluster-pulse service, artifact URLs from the taskcluster-queue
+service, and so on.
+
+The `rootUrl` is a single domain, with all Taskcluster services available at
+sub-paths such as `/api/queue`. Users of the current Taskcluster installation
+will note that this is a change: queue is currently at
+`https://queue.taskcluster.net`, not `https://taskcluster.net/queue`. We have
+solved this issue by special-casing the rootUrl `https://taskcluster.net` to
+generate the old-style URLs. Once we have migrated all users out of the current
+installation, we will remove that special-case.
+
+### Data Migrations
+
+The first few deployments of Taskcluster will not require great support for
+migrations. A staging environment, for example, can be completely destroyed and
+re-created without any adverse impact.  But we will soon need to support users
+upgrading Taskcluster from earlier releases with no (or at least minimal)
+downtime.
+
+Our Azure tables library
+([azure-entities](https://github.com/taskcluster/azure-entities)) already has
+rudimentary support for schema updates, so modifying the structure of table
+rows is not difficult, although refactoring a single table into multiple tables
+would be difficult.
+
+As we transition to using Postgres instead of Azure, we will need to adopt some
+of the common migration tools. Ideally we can support downtime-free upgrades
+like azure-entities does, instead of requiring downtime to run DB migrations
+synchronously. [Bug 1431783](https://bugzilla.mozilla.org/show_bug.cgi?id=1431783)
+tracks this work.
+
+### Customization
+
+As a former maintainer of Buildbot, I've had a lot of experience with CI
```

Comment:
Honestly you may want to mention this earlier. It lends a lot of credibility to anything you say in this post!

Reply:
I didn't want to overplay my hand :)